### PR TITLE
Ensure pods are updated if the service account is

### PR DESCRIPTION
**Pull request by idlewis**
This deals with #350

After this change, the pods will be deleted/recreated automatically if a pullSecret is added to the CR

Also, the pull secret will be deleted from the service account if it is deleted from the CR.

### CODE REVIEW

**Comment by idlewis** [comment]
@halim-lee This is the PR for issue #350

File: utils/utils.go [hunk 1]
```diff
@@ -591,6 +591,13 @@ func CustomizePodSpec(pts *corev1.PodTemplateSpec, ba common.BaseComponent) {
 		})
 	}
 
+	// This ensures that the pods are updated if the service account is updated
```
Comment (leochr) [inline utils/utils.go:594]: The pods should be updated for user-provided service account as well. Setting the reference value at the end of `ServiceAccountPullSecretExists` would ensure that the right service account is checked for existence, it has image pull secret (if it exists) and finally set the resource version of the corresponding service account.

File: utils/utils.go [hunk 2]
```diff
@@ -673,6 +686,15 @@ func CustomizeServiceAccount(sa *corev1.ServiceAccount, ba common.BaseComponent)
 	}
 }
 
+func removePullSecret(sa *corev1.ServiceAccount, pullSecretName string) {
+	for i, obj := range sa.ImagePullSecrets {
+		if obj.Name == pullSecretName {
+			sa.ImagePullSecrets = append(sa.ImagePullSecrets[:i], sa.ImagePullSecrets[i+1:]...)
```
Comment (leochr) [inline utils/utils.go:692]: I am not entirely sure if a concurrent modification exception could happen in Go for this scenario (iterating through pull secrets and setting a value for the same). If we are not sure, getting the index of the pull secret we are looking for and setting this outside of for-loop is another option.

File: utils/utils.go [hunk 3]
```diff
@@ -662,7 +669,13 @@ func CustomizeServiceAccount(sa *corev1.ServiceAccount, ba common.BaseComponent)
...
 	sa.Annotations = MergeMaps(sa.Annotations, ba.GetAnnotations())
 
+	psr := ba.GetStatus().GetReferences()[common.StatusReferencePullSecretName]
+	if psr != "" && (ba.GetPullSecret() == nil || *ba.GetPullSecret() != psr) {
+		removePullSecret(sa, psr)
+	}
+
 	if ba.GetPullSecret() != nil {
+		ba.GetStatus().SetReference(common.StatusReferencePullSecretName, *ba.GetPullSecret())
 		if len(sa.ImagePullSecrets) == 0 {
```
Comment (leochr) [inline utils/utils.go:679]: Adding the pull secret by iterating through the secrets would be better than adding it based on index. WSA for [reference](https://github.ibm.com/websphere/automation-operator/blob/ab96861893c32f30f9918ee8d939db7b7c6fcaa6/controllers/websphereautomation_controller.go#L543)

**Review by leochr — COMMENTED** [review]
@idlewis thanks for the PR and @halim-lee for the updates. Added some comments.

File: utils/utils.go [hunk 4]
```diff
@@ -662,7 +669,13 @@ func CustomizeServiceAccount(sa *corev1.ServiceAccount, ba common.BaseComponent)
 	sa.Labels = ba.GetLabels()
 	sa.Annotations = MergeMaps(sa.Annotations, ba.GetAnnotations())
 
+	psr := ba.GetStatus().GetReferences()[common.StatusReferencePullSecretName]
+	if psr != "" && (ba.GetPullSecret() == nil || *ba.GetPullSecret() != psr) {
+		removePullSecret(sa, psr)
```
Comment (leochr) [inline utils/utils.go:674]: @idlewis `common.StatusReferencePullSecretName` should also be unset once the pull secret is removed, so this if-condition is not continuously executed.

**Comment by idlewis** [comment]
closing in favour of PR #387 as I've tweaked the commits a fair bit. All the review comments from this PR are addressed in 387

**Pull request closed without merging**